File: other/largestRectangleArea.cpp
// leetcode 84. largestRectangleArea cpp

class Solution {
public:
    int largestRectangleArea(vector<int>& heights) 
    {
        heights.insert(heights.begin(), 0);
        heights.insert(heights.end(), 0);
        stack<int> stk;
        int ret = 0;
        for(int i = 0; i < heights.size(); i++)
        {
            while(!stk.empty() && heights[stk.top()] > heights[i])
            {
                int h = heights[stk.top()];
                stk.pop();
                ret = max(ret, h * (i - stk.top() - 1));
            }
            stk.push(i);
        }
        return ret;
    }
};

File: leetcode_hot_100/largestRectangleArea.py
# hot100-lc 84. largestRectangleArea

class Solution(object):
    def largestRectangleArea(self, heights):
        """
        :type heights: List[int]
        :rtype: int
        """
        ret = 0
        stack = []
        i = 0
        while i < len(heights):
            if len(stack) == 0:
                stack.append(i)
                i += 1
                continue
            if heights[i] > heights[stack[-1]]:
                stack.append(i)
                i += 1
            elif heights[i] < heights[stack[-1]]:
                left = stack[-1]
                while left >= 0:
                    if heights[left] < heights[stack[-1]]:
                        break
                    left -= 1
                ret = max(ret, heights[stack[-1]] * (i - left - 1))
                stack.pop()
            else:
                i += 1
                continue

        for pos in stack:
            left = pos
            h = heights[pos]
            while left >= 0:
                if heights[left] < h:
                    break
                left -= 1
            ret = max(ret, heights[pos] * (i - left - 1))
        return ret
